packages/iTwinUI-react/src/core/Table/Table.tsx

```diff
@@ -655,6 +655,12 @@ export const Table = <
     setPageSize(pageSize);
   }, [pageSize, setPageSize]);
 
+  // Reset the column order whenever new columns are passed
+  // This is to avoid the old columnOrder from affecting the new columns' columnOrder
+  React.useEffect(() => {
```

Comment:
To me this seems like a drastic action to reset order anytime new columns are passed.
I'm trying to think of a use case when user want to keep old order but nothing comes to my mind right now.

One idea might be to add prop for toggling reset/no-reset. Default would be reset.

But I'm not sure about that.

Reply:
> To me this seems like a drastic action to reset order anytime new columns are passed.

Yes, I too am a bit hesitant to doing this reset. Because if users do not memoize the columns they pass (using `React.memo` or similar), then the columns object could be re-created on each render.
e.g.
```tsx
const columns = () => [ ... ];

<Table
  columns={columns()}
  ...
/>
```

In this case, reordering columns may not work as the order is always reset.

---

> I'm trying to think of a use case when user want to keep old order but nothing comes to my mind right now.

I tried running the existing tests. All passed except one that was re-rendering with a new columns object (like the above example) (near [L3321-L3350](https://github.com/iTwin/iTwinUI-react/pull/983/files#diff-c4aa401f671f853692462c6d1ee006969f80e3d02863b8a1a8919154aa3f67d9R3321-R3350)). I edited that test to check that the column order is not `resultingColumns` but instead `columns` itself since the order is expected to reset.

---

> One idea might be to add prop for toggling reset/no-reset. Default would be reset.

Maybe we name it something like `autoResetColumnOrder`.

```tsx
/**
* If this is:
* - true, then the `columnOrder` is reset whenever the contents of the passed `columns` changes.
* - false, then `columnOrder` does not reset. This is useful when wanting to manually maintain columnOrder
* @default = true
*/
autoResetColumnOrder?: boolean
```

Maybe we also provide an optional `columnOrder` prop if they want to programmatically change the column order? It would make more sense to provide the above prop if users can actually handle the `columnOrder` themselves.

One other easier option could be to internally check if the content of `columns` have changed and only then do the reset. That way even the above example works.

What do you think?

Reply:
We need to be careful with adding props. Too many props are not helpful for the user. In this case, I think we can reset based on some if condition rather than on every render.

Reply:
@rohan-kadkol Check if columns actually changed in this useEffect.

Reply:
> One other easier option could be to internally check if the content of columns have changed and only then do the reset. That way even the above example with un-memomized columns works.

> @rohan-kadkol Check if columns actually changed in this useEffect.

**[DONE]** I now added a check to reset the `columnOrder` only when the "contents" of `columns` change ([L687](https://github.com/iTwin/iTwinUI-react/pull/983/files#diff-9edd44b431060c27e0260d286ee188bdf901225b1ee84ff6540f9d796dd0f999R687)). That way, even if a new `columns` object with the same contents is passed on each render, the `columnOrder` is not reset.